app/controllers/admin/orders_controller.rb

```diff
@@ -44,22 +49,77 @@ def destroy
     end
   end
 
+  def trash
+    @count_orders_deleted_by_status = count_orders_deleted_by_status
+    @pagy, @trash_orders = if username.present?
+                             pagy Order.only_deleted.filter_by_name(username),
+                                  items: Settings.length.per_page_5
+                           elsif status.present?
+                             pagy Order.only_deleted.send(status),
+                                  items: Settings.length.per_page_5
+                           else
+                             orders_deleted
+                           end
+  end
+
+  def delete
+    if @order.destroy
+      respond_to do |format|
+        format.js
+      end
+    else
+      flash[:danger] = t ".delete_fail"
+      redirect_to trash_admin_orders_path
+    end
+  end
+
+  def restore
+    if @order.restore recursive: true
+      respond_to do |format|
+        format.js
+      end
+    else
+      flash[:danger] = t ".restore_fail"
+      redirect_to trash_admin_orders_path
+    end
+  end
+
   private
 
   def check_valid_status_for_update?
-    return if Order.statuses.values.include?(status_params.to_i) &&
-              (@order.status_before_type_cast != status_params.to_i) &&
-              (Order.statuses[:pending] != status_params.to_i)
+    return if valid_status_for_update?
 
     flash[:danger] = t ".update_status_fail"
     redirect_to admin_orders_path
   end
 
-  def check_valid_status_for_destroy?
+  def valid_status_for_update?
+    Order.statuses.values.include?(status_params.to_i) &&
+      (@order.status_before_type_cast != status_params.to_i) &&
+      (Order.statuses[:pending] != status_params.to_i) &&
+      (@order.rejected? != true) &&
+      (@order.canceled? != true)
```

Comment:
```suggestion
      !@order.canceled?
```
dungf nhuw nayf ddc k a